### PR TITLE
Add Contract.creator and address checking primitives

### PR DESCRIPTION
**Pull request by hanssv**
Added Address.is_contract, Address.is_oracle, Oracle.check, Oracle.check_query
and Contract.creator to Sophia and AEVM

https://www.pivotaltracker.com/story/show/165440601
and https://www.pivotaltracker.com/story/show/165713319 

Doc: https://github.com/aeternity/protocol/pull/347

### CODE REVIEW

File: apps/aecore/src/aec_vm_chain.erl [hunk 1]
```diff
@@ -470,6 +512,16 @@ oracle_typerep(?ABI_AEVM_SOPHIA_1, BinaryFormat) ->
         _:_ -> {error, bad_typerep}
     end.
 
+%%    Address
+addr_is_contract(Addr, #state{ trees = CTrees }) ->
+    Trees = get_on_chain_trees(CTrees),
+    io:format("Adddr: ~p\n", [Addr]),
```
Comment (gorillainduction) [inline apps/aecore/src/aec_vm_chain.erl]: Stray io:format

Comment (hanssv) [reply apps/aecore/src/aec_vm_chain.erl]: Fixed.

File: apps/aevm/src/aevm_eeevm.erl [hunk 2]
```diff
@@ -425,6 +425,14 @@ loop(CP, StateIn) ->
                     <<Val:256/integer-unsigned>> = Hash,
                     State4 = push(Val, State3),
                     next_instruction(CP, State, State4);
+                %% 2f: Non-standard Env information
+                ?CREATOR ->
+                    %% 0x2f Creator δ=0 α=1
+                    %% Get address of contract creator.
+                    %% µ's[0] ≡ Ia
```
Comment (lucafavatella) [inline apps/aevm/src/aevm_eeevm.erl]: If so, what is difference with `0x30` ?

File: test/contracts/aevm_3/acm_dutch_auction.aes [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      height = Chain.block_height,
+      dec    = decrease,
+      sold   = false }
+
+  public stateful function bid() =
+    require( !(state.sold), "sold")
+    let price =
+      state.amount - (Chain.block_height - state.height) * state.dec
+    require( Contract.balance >= price, "no money")
+    Chain.spend(Contract.creator, price)
```
Comment (lucafavatella) [inline test/contracts/aevm_3/acm_dutch_auction.aes]: I see this is correctly copy of /test/contracts/acm_dutch_auction.aes , adapted. Ok.

File: apps/aevm/src/aevm_eeevm.erl (hunk 2 above)
Comment (hanssv) [reply apps/aevm/src/aevm_eeevm.erl]: This is the account that created the contract, not the contract itself.

File: rebar.lock [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {"1.1.0",
 [{<<"aebytecode">>,
   {git,"https://github.com/aeternity/aebytecode.git",
-      {ref,"2555868990ac2a08876e86b1b798b4750273591f"}},
+      {ref,"08a09b065b9c5c4d1c499372a90914c80349dace"}},
```
Comment (lucafavatella) [inline rebar.lock]: I see https://github.com/aeternity/aebytecode/compare/2555868990ac2a08876e86b1b798b4750273591f...08a09b065b9c5c4d1c499372a90914c80349dace

File: rebar.config [hunk 5]
```diff
@@ -183,7 +183,7 @@
                                  {sname, 'aeternity_ct@localhost'}]},
                     {deps, [{meck, "0.8.12"},
                             {websocket_client, {git, "git://github.com/aeternity/websocket_client", {ref, "a4fb3db"}}},
-                            {aesophia, {git, "https://github.com/aeternity/aesophia.git", {ref,"0aa1c89"}}},
+                            {aesophia, {git, "https://github.com/aeternity/aesophia.git", {ref,"f8ee8f7"}}},
```
Comment (lucafavatella) [inline rebar.config]: I see https://github.com/aeternity/aesophia/compare/0aa1c89...f8ee8f7

File: apps/aevm/src/aevm_eeevm.erl (hunk 2 above)
Comment (lucafavatella) [reply apps/aevm/src/aevm_eeevm.erl]: What I mean is that this line does not define the semantics of the instruction, and is rather confusing. I suggest removing it.
```suggestion
```

File: apps/aevm/src/aevm_eeevm.erl [hunk 6]
```diff
@@ -425,6 +425,13 @@ loop(CP, StateIn) ->
                     <<Val:256/integer-unsigned>> = Hash,
                     State4 = push(Val, State3),
                     next_instruction(CP, State, State4);
+                %% 2f: Non-standard Env information
+                ?CREATOR ->
+                    %% 0x2f Creator δ=0 α=1
+                    %% Get address of contract creator.
```
Comment (lucafavatella) [inline apps/aevm/src/aevm_eeevm.erl]: So basically the intention is exposing the `owner` field [already stored in the contract](https://github.com/aeternity/protocol/blame/c638d7728380bfdf85968dba1d22672f49664727/serializations.md#L387)?

Comment (hanssv) [reply apps/aevm/src/aevm_eeevm.erl]: Yes, and this instruction was already advertised in the Sophia documentation, but we forgot to implement it ;-)

File: test/contracts/aevm_3/acm_dutch_auction.aes [hunk 7]
```diff
@@ -0,0 +1,25 @@
+contract DutchAuction =
```
Comment (lucafavatella) [inline test/contracts/aevm_3/acm_dutch_auction.aes]: Could you please point me to where this file is hooked to CI?

File: test/contracts/address_checks.aes [hunk 8]
```diff
@@ -0,0 +1,41 @@
+contract AddrCheck =
```
Comment (lucafavatella) [inline test/contracts/address_checks.aes]: I see this hooked in `aecontract_SUITE` - ok.

File: apps/aecore/src/aec_vm_chain.erl [hunk 9]
```diff
@@ -459,6 +463,44 @@ oracle_response_format(Oracle, #state{trees = ChainTrees}) ->
...
+            ABIVersion         = aeo_oracles:abi_version(O),
+            case ABIVersion of
+                ?ABI_NO_VM -> %% Question/Response is non-sophia string
+                    Equal = string == QFormat andalso string == ChainQFormat andalso
+                            string == RFormat andalso string == ChainRFormat,
+                    {ok, bool2word(Equal)};
+                _Other ->
+                    Equal = (QFormat == ChainQFormat) andalso (RFormat == ChainRFormat),
+                    {ok, bool2word(Equal)}
+            end;
```
Comment (UlfNorell) [inline apps/aecore/src/aec_vm_chain.erl]: `ABIVersion` check should not be necessary, and the whole block can be replace with a call to `oracle_check`.

File: apps/aevm/src/aevm_chain_api.erl [hunk 10]
```diff
@@ -135,6 +135,26 @@
                                     ChainState :: chain_state()) ->
     {ok, aeo_oracles:relative_ttl()} | {error, term()}.
 
+-callback oracle_check(Oracle :: pubkey(),
+                       RFormat :: aeb_aevm_data:type(),
+                       QFormat :: aeb_aevm_data:type(),
```
Comment (UlfNorell) [inline apps/aevm/src/aevm_chain_api.erl]: Wrong argument order.

File: apps/aevm/src/aevm_chain_api.erl [hunk 11]
```diff
@@ -135,6 +135,26 @@
...
+                       ChainState :: chain_state()) ->
+    {ok, non_neg_integer()} | {error, term()}.
+
+-callback oracle_check_query(Oracle :: pubkey(),
+                             Query :: pubkey(),
+                             RFormat :: aeb_aevm_data:type(),
+                             QFormat :: aeb_aevm_data:type(),
```
Comment (UlfNorell) [inline apps/aevm/src/aevm_chain_api.erl]: .. and here

File: test/contracts/aevm_3/acm_dutch_auction.aes (hunk 7 above)
Comment (hanssv) [reply test/contracts/aevm_3/acm_dutch_auction.aes]: `aehttp_contract_SUITE`

File: apps/aecontract/test/aecontract_SUITE.erl [hunk 12]
```diff
@@ -4083,6 +4085,67 @@ sophia_bytes(_Cfg) ->
...
+    [ ?assertEqual(false, ?call(call_contract, Acc, Cx, check_oq1, bool, {C2, X}))
+      || Cx <- [C1, C2], X <- [<<OQ21:256>>, <<OQ22:256>>] ],
+
+    %% Check that queries from oracle 1 does pass as query 1
+    [ ?assertEqual(true, ?call(call_contract, Acc, Cx, check_oq1, bool, {C1, X}))
+      || Cx <- [C1, C2], X <- [<<OQ11:256>>, <<OQ12:256>>] ],
+
+    %% Check that queries from oracle 2 does pass as query 2
+    [ ?assertEqual(true, ?call(call_contract, Acc, Cx, check_oq2, bool, {C2, X}))
+      || Cx <- [C1, C2], X <- [<<OQ21:256>>, <<OQ22:256>>] ],
```
Comment (UlfNorell) [inline apps/aecontract/test/aecontract_SUITE.erl]: Maybe test that queries for one oracle doesn't pass as queries for another, even if they have the same type.

File: test/contracts/aevm_3/acm_dutch_auction.aes (hunk 7 above)
Comment (lucafavatella) [reply test/contracts/aevm_3/acm_dutch_auction.aes]: https://github.com/aeternity/aeternity/pull/2398/commits/c2400163f4ce453bc47d687d4f7f5844e37065bc ?
Thanks.

File: apps/aehttp/test/aehttp_contracts_SUITE.erl [hunk 13]
```diff
@@ -1020,7 +1020,10 @@ acm_dutch_auction_contract(Config) ->
                  priv_key := CPriv}} = proplists:get_value(accounts, Config),
 
     %% Compile test contract
-    Contract = compile_test_contract("acm_dutch_auction"),
+    Contract = case aect_test_utils:latest_protocol_version() of
+                   Vsn when Vsn < ?FORTUNA_PROTOCOL_VSN -> compile_test_contract("acm_dutch_auction");
+                   _                                    -> compile_test_contract("aevm_3/acm_dutch_auction")
```
Comment (lucafavatella) [inline apps/aehttp/test/aehttp_contracts_SUITE.erl]: I see test hooked. Ok.

**Pull request merged**